runtime/wasm-bpf-rs/src/bpf/poll.rs

```diff
@@ -130,167 +42,174 @@ pub fn wasm_bpf_buffer_poll(
+        poller.result_container.clone()
+    };
+    // Here we could try to extract the result
+    if let Some(v) = result_container.borrow_mut().take() {
+        let memory = match caller.get_memory() {
+            Err(e) => {
+                error!("Failed to get exported memory: {}", e);
+                return -1;
             }
+            Ok(v) => v,
         };
-        if inner.inner_ptr().is_null() {
+        let bytes_to_write = v.len().min(max_size as usize);
+        if let Err(e) = memory.write(&mut caller, data as usize, &v[..bytes_to_write]) {
+            error!("Failed to write wasm memory: {}", e);
             return -1;
         }
-        self.inner = inner;
-        0
-    }
-
-    /// polling the bpf buffer
-    pub fn bpf_buffer__poll(&self, timeout_ms: i32) -> i32 {
-        match self.inner {
-            BufferInnerType::PerfBuf(s) => unsafe { perf_buffer__poll(s, timeout_ms) },
-            BufferInnerType::RingBuffer(s) => unsafe { ring_buffer__poll(s, timeout_ms) },
-            BufferInnerType::Unsupported => -EINVAL,
-        }
-    }
-}
-
-impl Drop for BpfBuffer {
-    fn drop(&mut self) {
-        match self.inner {
-            BufferInnerType::PerfBuf(s) => unsafe {
-                perf_buffer__free(s);
-            },
-            BufferInnerType::RingBuffer(s) => unsafe { ring_buffer__free(s) },
-            BufferInnerType::Unsupported => {}
+        // Call the callback
+        if caller.data().wrapper_called {
+            let mut result = [Val::I32(0)];
+            let callback = caller.data().callback_func_name.clone();
+            if let Err(err) = caller
+                .get_export(&callback)
+                .unwrap()
+                .into_func()
+                .unwrap()
+                .call(
+                    &mut caller,
+                    &[
+                        // Seems that tinygo cannot produce unsigned integer types, so just let wasmtiime to perform the conversion
+                        Val::I32(ctx as i32),
+                        Val::I32(data as i32),
+                        Val::I32(bytes_to_write as i32),
+                    ],
+                    &mut result,
+                )
+            {
+                error!("Failed to call the callback through direct export: {}", err);
+                return -1;
+            }
+        } else {
+            match caller.perform_indirect_call::<SampleCallbackParams, SampleCallbackReturn>(
+                sample_func,
+                (ctx, data, bytes_to_write as u32),
+            ) {
+                Ok(v) => {
+                    return v;
+                }
+                Err(e) => {
+                    error!(
+                        "Failed to perform indirect call when polling: {} ; {}\n{}",
+                        e.to_string(),
+                        e.root_cause(),
+                        e.backtrace()
+                    );
+                    return 0;
+                }
+            }
```

Comment:
It may be possible to use `map_err`, I simply tried this modification
```rs
    let _ = caller
        .perform_indirect_call::<SampleCallbackParams, SampleCallbackReturn>(
            sample_func,
            (ctx, data, bytes_to_write as u32),
        )
        .map_err(|e| {
            error!(
                "Failed to perform indirect call when polling: {} ; {}\n{}",
                e.to_string(),
                e.root_cause(),
                e.backtrace()
            );
            return 0;
        });
```